testsuite/cluster-test/src/atomic_histogram.rs

```diff
@@ -0,0 +1,230 @@
+        );
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.buckets.len() {
+            buf.push(self.buckets[i] - other.buckets[i]);
+        }
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_millis: self.step_millis,
+            buckets: buf,
+        }
+    }
+}
+
+impl AtomicHistogramSnapshot {
+    pub fn cal_percentile_latency(
+        &self,
+        numerator: u64,
+        denominator: u64,
+        committed: u64,
+        snapshot: &AtomicHistogramSnapshot,
+    ) -> u64 {
+        let p_count = committed * numerator / denominator;
+        let mut counter = 0u64;
+        let mut total_latency = 0u64;
+        for i in 0..snapshot.buckets.len() {
+            let num = snapshot.buckets[i];
+            let gap = p_count - counter;
+            if counter >= p_count {
+                break;
+            }
+            if num > gap {
+                total_latency += gap * self.step_millis * i as u64;
+                counter += gap;
+            } else {
+                total_latency += num * self.step_millis * i as u64;
+                counter += num;
+            }
+        }
+        if p_count == 0 {
+            return 0;
+        }
+        total_latency / p_count
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::tx_emitter::*;
+    // other use:: for test only
+
+    #[test]
+    pub fn test_default_atomic_histogram() {
+        let histogram = AtomicHistogramAccumulator::default();
+        assert_eq!(histogram.step_millis, DEFAULT_HISTOGRAM_STEP_MILLIS);
+        assert_eq!(histogram.buckets.len(), DEFAULT_HISTOGRAM_CAPACITY);
+    }
+
+    #[test]
+    pub fn test_get_bucket_num() {
+        let histogram = AtomicHistogramAccumulator::default();
+        assert_eq!(histogram.get_bucket_num(0), 0);
+        assert_eq!(histogram.get_bucket_num(49), 0);
+        assert_eq!(histogram.get_bucket_num(50), 1);
+        assert_eq!(histogram.get_bucket_num(51), 1);
+        assert_eq!(histogram.get_bucket_num(200_000), 2047);
+    }
+
+    #[test]
+    pub fn test_sub() {
+        let mut cur_snap = AtomicHistogramSnapshot::default();
+        let mut cur_vec = Vec::new();
+        for i in 10..20 {
+            cur_vec.push(i);
+        }
+        cur_snap.buckets = cur_vec;
+
+        let mut pre_snap = AtomicHistogramSnapshot::default();
+        let mut prev_vec = Vec::new();
+        for i in 0..10 {
+            prev_vec.push(i);
+        }
+        pre_snap.buckets = prev_vec;
+        let res = &cur_snap - &pre_snap;
+        for &i in res.buckets.iter() {
+            assert_eq!(i, 10);
+        }
+    }
+
+    #[test]
+    pub fn test_percentile_latency() {
+        let histogram = AtomicHistogramAccumulator::default();
+        // set 10 commits, with latency as:
+        // 100ms, 200ms, 300ms ... 1000ms
+        // for p90 count is 9
+        // expected p90 = (100 + 200 + ... + 900) / 9 = 500
```

Comment:
So this is not what p90 is actually :)
For this test case, p90 is 900ms:
90% of transactions has faster or equal top90
10% of transactions are slower then p90